### PR TITLE
2 add training endpoints

### PR DESCRIPTION
**Pull request by ao-picterra**
- Add Training API functions
- Add above functions test
- Add full example
- Fix uploadRaster function

### CODE REVIEW

File: dist/Client.js [hunk 1]
```diff
@@ -146,23 +181,28 @@ class APIClient {
 
 
   async uploadRaster(fileName, rasterName) {
+    // Compute file size
     const stream = createReadStream(fileName);
-    let response, data;
-    response = await this._request( // Get upload URL
-    '/rasters/upload/file/', 'POST', {
+    const stats = statSync(fileName);
+    const fileSizeInBytes = stats['size'];
```
Comment (julienr) [inline dist/Client.js:187]: Do we need this change ? Upload is working fine for me on master, I feel like fetch is doing this already automatically when passed a stream. Or maybe I'm missing what this is fixing ?

File: dist/Client.js [hunk 2]
```diff
@@ -37,14 +38,22 @@ class APIError extends Error {
     this.body = body;
   }
 
+}
+
+exports.APIError = APIError;
+
+class ValidationError extends Error {
```
Comment (julienr) [inline dist/Client.js:45]: I feel I wouldn't introduce validation of detector settings or annotation types. Those are things that are validated by our backend, so if the user put something wrong, I feel we should just surface the backend error.

The issue with validating those here is that if they change, you have to update the validation logic. So I'd keep this simple and rely on backend validation as much as possible.

File: dist/Client.js [hunk 3]
```diff
@@ -316,6 +356,116 @@ class APIClient {
...
+
+
+    resp = await this._request(`/detectors/${detectorId}/training_rasters/${rasterId}/${annotationType}/upload/bulk/`, 'POST');
+    await checkResponse(resp);
+    data = await resp.json();
+    const uploadUrl = data['upload_url'];
+    const uploadId = data['upload_id']; // Put data in remote storage
+
+    resp = await this._request(uploadUrl, 'PUT', {
+      /* 'Content-Length': fileSizeInBytes, */
```
Comment (julienr) [inline dist/Client.js:440]: we can chop this comment

File: dist/Client.js [hunk 4]
```diff
@@ -316,6 +356,116 @@ class APIClient {
...
+      const validTypes = annotationTypes.join(', ');
+      throw new ValidationError(`Invalid annotation type ${annotationType}; allowed values: ${validTypes}.`);
+    } // Get upload URL
+
+
+    resp = await this._request(`/detectors/${detectorId}/training_rasters/${rasterId}/${annotationType}/upload/bulk/`, 'POST');
+    await checkResponse(resp);
+    data = await resp.json();
+    const uploadUrl = data['upload_url'];
+    const uploadId = data['upload_id']; // Put data in remote storage
```
Comment (julienr) [inline dist/Client.js:437]: comment seems outdated

File: dist/Client.js [hunk 5]
```diff
@@ -316,6 +356,116 @@ class APIClient {
...
+
+    resp = await this._request(uploadUrl, 'PUT', {
+      /* 'Content-Length': fileSizeInBytes, */
+      'Content-Type': 'application/json'
+    }, JSON.stringify(annotationsGeoJSon), false);
+    await checkResponse(resp); // Start commit
+
+    resp = await this._request(`/detectors/${detectorId}/training_rasters/${rasterId}/${annotationType}/upload/bulk/${uploadId}/commit/`, 'POST');
+    await checkResponse(resp);
+    data = await resp.json(); // Poll until operation finishes
```
Comment (julienr) [inline dist/Client.js:447]: comment seems outdated

File: dist/Client.js [hunk 6]
```diff
@@ -316,6 +356,116 @@ class APIClient {
...
+   * @param {String} detectorId Identifier for the detector
+   * @throws {APIError} Containing error code and text
+   */
+
+
+  async trainDetector(detectorId) {
+    // Launch training
+    const response = await this._request(`/detectors/${detectorId}/train/`, 'POST');
+    await checkResponse(response);
+    const data = await response.json(); // Poll until training finishes
```
Comment (julienr) [inline dist/Client.js:465]: comment should be on next line ? But generally, I feel like this doesn't add much value given the next line is doing a 'wait', should be pretty obvious we're waiting for the operation to finish

**Review by julienr — CHANGES_REQUESTED** [review]
Thanks !

File: dist/Client.js (hunk 1 above)
Comment (ao-picterra) [reply dist/Client.js:187]: I've tried the `training.js` example code and UploadRaster raises a 411: which node-fetch version do you have?

Comment (julienr) [reply dist/Client.js:187]: Mmmh works for me, both on node-fetch 2.6.1 and 2.6.0 (which is our min requirements). Are you sure it's not an issue that the file is non-existing or something ?

Comment (ao-picterra) [reply dist/Client.js]: My bad, messed up with npm, works...

File: dist/Client.js (hunk 3 above)
Comment (ao-picterra) [reply dist/Client.js]: Checked

File: dist/Client.js (hunk 4 above)
Comment (ao-picterra) [reply dist/Client.js]: Checked

File: dist/Client.js (hunk 5 above)
Comment (ao-picterra) [reply dist/Client.js]: Checked

File: dist/Client.js (hunk 6 above)
Comment (ao-picterra) [reply dist/Client.js]: Checked

File: dist/Client.js (hunk 1 above)
Comment (ao-picterra) [reply dist/Client.js]: It was not npm though, but the fact that I was testing locally: apparently minio raises a 411 if not PUTting with content-length, I'll investigate (not smth urgent anyway)

Comment (julienr) [reply dist/Client.js]: Ah ok its a minio / gcloud difference. Mmmh ok then maybe we can add it back at some point to be able to test.

**Review by julienr — APPROVED** [review]

**Pull request merged**